src/utils/rooms.ts

```diff
@@ -24,3 +36,138 @@ export function privateShouldBeEncrypted(): boolean {
+        action: Action.ViewRoom,
+        auto_join: autoJoin,
+        should_peek: shouldPeek,
+        metricsTrigger: "RoomDirectory",
+    };
+    if (room) {
+        // Don't let the user view a room they won't be able to either
+        // peek or join: fail earlier so they don't have to click back
+        // to the directory.
+        if (client.isGuest()) {
+            if (!room.world_readable && !room.guest_can_join) {
+                dis.dispatch({ action: 'require_registration' });
+                return;
+            }
+        }
+
+        if (!roomAlias) {
+            roomAlias = getDisplayAliasForAliasSet(room.canonical_alias, room.aliases);
+        }
+
+        payload.oob_data = {
+            avatarUrl: room.avatar_url,
+            // XXX: This logic is duplicated from the JS SDK which
+            // would normally decide what the name is.
+            name: room.name || roomAlias || _t('Unnamed room'),
+        };
+
+        if (roomServer) {
+            payload.via_servers = [roomServer];
+        }
+    }
+    // It's not really possible to join Matrix rooms by ID because the HS has no way to know
+    // which servers to start querying. However, there's no other way to join rooms in
+    // this list without aliases at present, so if roomAlias isn't set here we have no
+    // choice but to supply the ID.
+    if (roomAlias) {
+        payload.room_alias = roomAlias;
+    } else {
+        payload.room_id = room.room_id;
+    }
+    dis.dispatch(payload);
+};
+
+interface IJoinRoomByAliasOpts {
+    instanceId?: string;
+    roomServer?: string;
+    protocols: Protocols;
+}
+
+export function joinRoomByAlias(cli: MatrixClient, alias: string, {
+    instanceId,
+    roomServer,
+    protocols,
+}: IJoinRoomByAliasOpts): void {
+    // If we don't have a particular instance id selected, just show that rooms alias
+    if (!instanceId || instanceId === ALL_ROOMS) {
+        // If the user specified an alias without a domain, add on whichever server is selected
+        // in the dropdown
+        if (alias.indexOf(':') == -1) {
+            alias = alias + ':' + roomServer;
+        }
+        showRoom(cli, null, {
+            roomAlias: alias,
+            autoJoin: true,
+        });
+    } else {
+        // This is a 3rd party protocol. Let's see if we can join it
+        const protocolName = protocolNameForInstanceId(protocols, instanceId);
+        const instance = instanceForInstanceId(protocols, instanceId);
+        const fields = protocolName
+            ? getFieldsForThirdPartyLocation(alias, protocols[protocolName], instance)
+            : null;
+        if (!fields) {
+            const brand = SdkConfig.get().brand;
+            Modal.createTrackedDialog('Unable to join network', '', ErrorDialog, {
+                title: _t('Unable to join network'),
+                description: _t('%(brand)s does not know how to join a room on this network', { brand }),
+            });
```

Comment:
the other approach for avoiding an import of `ErrorDialog` would be to move calls like this out to the `DialogOpener`